pepper-apis/src/main/java/org/broadinstitute/ddp/export/DataExporter.java

```diff
@@ -969,12 +973,13 @@ private String formatParticipantToStructuredJSON(
         return questionRecords.stream().filter(Objects::nonNull).collect(Collectors.toList());
     }
 
-    private QuestionRecord createRecordForComponent(String component, Participant participant) {
+    private QuestionRecord createRecordForComponent(Handle handle, FormResponse formResponse, String component, Participant participant) {
         ComponentQuestionRecord record = null;
         List<List<String>> values;
         switch (component) {
             case "MAILING_ADDRESS":
-                MailAddress address = participant.getUser().getAddress();
+                MailAddress address =
+                        getSnapshottedAddress(handle, formResponse.getId(), addressService, participant.getUser().getAddress());
```

Comment:
I think we have already queried for the activity substitutions data so doing another query (for each user and activity) is not as efficient. If you look at the `Participant` class, you'll see this field:
```
private Map<Long, Map<String, String>> activityInstanceSubstitutions;
```
And this is populated in `ExportUtil.extractParticipantsFromResultSet()`. So we have already bulk-queried for that data. Now, if only we can figure out a way to bulk query for the snapshotted addresses beforehand and pass it in, that'll be nice.

Reply:
> And this is populated in ExportUtil.extractParticipantsFromResultSet()
 Is this method called for all 4 cases?
 ```
 DataExporter.exportToElasticsearch()
   formatParticipantToStructuredJSON()
   formatParticipantToFlatJSON
DataExporter.exportDataSetAsCsv()
CustomExporter.exportDataSetAsCsv()
```

Reply:
Reused Participant#activityInstanceSubstitutions when it is possible